### PR TITLE
 feat(sync-actions): add support for custom types

### LINKED ISSUE #679: [sync-actions]: support sync action on custome types
### Support reference number

SUPPORT-3365

### Description

Create a sync action that allows to build API update actions for [Types](https://docs.commercetools.com/http-api-projects-types.html#types)

### PR DESCRIPTION
**Pull request by daern91**
#### Summary

PR adds support for the [types API](https://docs.commercetools.com/http-api-projects-types.html) to the sync-actions package.

It includes all available update actions and follows the structure for `product-types` that got introduced here https://github.com/commercetools/nodejs/commit/17be8117e822a9d47d31abd686c5c37093c1f539

#### Todo

- Tests
  - [x] Unit
- [x] Documentation

resolves #679

### CODE REVIEW

File: docs/sdk/api/syncActions.md [hunk 1]
```diff
@@ -402,11 +402,11 @@ client.execute(taxCategoriesRequests)
 
 > From package [@commercetools/sync-actions](/sdk/api/README.md#sync-actions).
 
-Creates a [sync action](/sdk/Glossary.md#sync-action) that allows to build API update actions for _zones_.
+Creates a [sync action](/sdk/Glossary.md#syncaction) that allows to build API update actions for _zones_.
 
 #### Arguments
 
-1.  `actionGroups` _(Array)_: A list of [action group](/sdk/Glossary.md#sync-action) in case some actions need to be _blacklisted_ or _whitelisted_.
+1.  `actionGroups` _(Array)_: A list of [action group](/sdk/Glossary.md#syncaction) in case some actions need to be _blacklisted_ or _whitelisted_.
```
Comment (adnasa) [inline docs/sdk/api/syncActions.md]: I don't see how these changes are related.. 😞

File: packages/sync-actions/test/types-sync-base.spec.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    beforeEach(() => {
+      before = {
+        key: 'orwell-key',
+      }
+      now = {
+        key: 'huxley-key',
+      }
+      updateActions = typesSync.buildActions(now, before)
+    })
+    test('should return `changeKey` update-action', () => {
```
Comment (adnasa) [inline packages/sync-actions/test/types-sync-base.spec.js]: I would like to see test `null -> ''` or `'' -> null`, given shouldOmitEmptyString

File: packages/sync-actions/src/types-actions.js [hunk 3]
```diff
@@ -15,3 +24,55 @@ export function actionsMapBase(diff, oldObj, newObj, config = {}) {
...
+      } else if (diffValue.label) {
+        actions.push({
+          action: 'changeLabel',
+          label: extractedPairs.newObj.label,
+          fieldName: extractedPairs.oldObj.name,
+        })
+      }
+    } else if (getIsRemovedOperation(diffKey)) {
+      if (Array.isArray(diffValue)) {
+        if (diffValue.length === 3 && diffValue[2] === 3) {
```
Comment (adnasa) [inline packages/sync-actions/src/types-actions.js]: nothing actionable, but I just want to rant that these magic numbers are driving me nuts

File: packages/sync-actions/src/types.js [hunk 4]
```diff
@@ -4,15 +4,29 @@ import createBuildActions from './utils/create-build-actions'
 import createMapActionGroup from './utils/create-map-action-group'
 import * as typeActions from './types-actions'
 import * as diffpatcher from './utils/diffpatcher'
+import findMatchingPairs from './utils/find-matching-pairs'
 
-const actionGroups = ['base', 'attributes']
+const actionGroups = ['base', 'fieldDefinitions']
```
Comment (adnasa) [inline packages/sync-actions/src/types.js]: have you reached out to the teams using this sync-action?

File: packages/sync-actions/test/types-sync-fields.spec.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+            name: 'text-field-definition',
+            label: {
+              en: 'text-en',
+            },
+          },
+        ],
+      })
+      updateActions = typesSync.buildActions(now, before)
+    })
+    test('should return `removeFieldDefinition` and `addFieldDefinition` actions', () => {
```
Comment (adnasa) [inline packages/sync-actions/test/types-sync-fields.spec.js]: I would like to see a test with the following data:

```js
previous = [
  {
    // no change
  },
  {
    // to be removed
  },
  {
    // will change to 'z'
    name: 'x'
  },
]
next = [
  {
    // no change
  },
  {
    // changed from 'x'
    name: 'z'
  },
]
````

File: docs/sdk/api/syncActions.md (hunk 1 above)
Comment (daern91) [reply docs/sdk/api/syncActions.md]: Just realised they had been wrong all along so included a little fix for it.

File: packages/sync-actions/src/types.js (hunk 4 above)
Comment (adnasa) [reply packages/sync-actions/src/types.js]: nevermind... we do not administer the types in our UI

File: packages/sync-actions/test/types-sync-fields.spec.js (hunk 5 above)
Comment (daern91) [reply packages/sync-actions/test/types-sync-fields.spec.js]: There is no update action for changing the `name`, only the `label`([ref](https://docs.commercetools.com/http-api-projects-types.html#change-fielddefinition-label)). 

Anyways, added that test here https://github.com/commercetools/nodejs/pull/757/commits/6b3da242b3d8d38f5ea7596781f3c6ea4f230bcb

Comment (daern91) [reply packages/sync-actions/test/types-sync-fields.spec.js]: @adnasa added here https://github.com/commercetools/nodejs/pull/757/commits/f369ff0feb5431701dc4ddaea61ecdd47068f934

File: packages/sync-actions/test/types-sync-fields.spec.js [hunk 6]
```diff
@@ -325,14 +325,14 @@ describe('Actions', () => {
     })
     test('should return `changeLabel` and `removeFieldDefinition` action', () => {
       expect(updateActions).toEqual([
+        { action: 'removeFieldDefinition', fieldName: 'should-be-removed' },
```
Comment (adnasa) [inline packages/sync-actions/test/types-sync-fields.spec.js]: 🔥

**Pull request merged**